plugins/processors/awsappsignals/internal/cardinalitycontrol/metrics_limiter.go

```diff
@@ -0,0 +1,414 @@
+	m.mapLock.RUnlock()
+	if svc == nil {
+		m.mapLock.Lock()
+		svc = m.services[serviceName]
+		if svc == nil {
+			svc = newService(serviceName, m.DropThreshold, m.RotationInterval, m.ctx, m.logger)
+			m.services[serviceName] = svc
+		}
+		m.mapLock.Unlock()
+	}
+
+	metricData := newMetricData(serviceName, metricName, labels)
+
+	reserved, _ := attributes.Get(common.AttributeTmpReserved)
+	if reserved.Bool() {
+		attributes.Remove(common.AttributeTmpReserved)
+		return true, nil
+	}
+
+	if !svc.admitMetricData(metricData) {
+		svc.rollupMetricData(attributes)
+
+		svc.totalRollup++
+		admitted = false
+
+		if m.LogDroppedMetrics {
+			m.logger.Debug(fmt.Sprintf("[%s] drop metric data", svc.name), zap.Any("labels", metricData.labels))
+		}
+	}
+
+	svc.totalMetricSent++
+
+	svc.rwLock.RLock()
+	defer svc.rwLock.RUnlock()
+
+	svc.totalCount++
+	svc.InsertMetricDataToPrimary(metricData)
+	svc.InsertMetricDataToSecondary(metricData)
+	return admitted, nil
+}
+
+func (m *MetricsLimiter) filterAWSDeclaredAttributes(attributes, resourceAttributes pcommon.Map) (map[string]string, string, bool) {
+	svcNameAttr, exists := attributes.Get(common.MetricAttributeLocalService)
+	if !exists {
+		return nil, "", false
+	}
+	labels := map[string]string{}
+	svcName := svcNameAttr.AsString()
+	for _, attrKey := range awsDeclaredMetricAttributes {
+		if attr, ok := attributes.Get(attrKey); ok {
+			labels[attrKey] = attr.AsString()
+		}
+	}
+	return labels, svcName, true
+}
+
+func (m *MetricsLimiter) removeStaleServices() {
+	var svcToRemove []string
+	for name, svc := range m.services {
+		if svc.rotations > 3 {
+			if svc.countSnapshot[0] == svc.countSnapshot[1] && svc.countSnapshot[1] == svc.countSnapshot[2] {
+				svc.cancelFunc()
+				svcToRemove = append(svcToRemove, name)
+			}
+		}
+	}
+
+	m.mapLock.Lock()
+	defer m.mapLock.Unlock()
+
+	for _, name := range svcToRemove {
+		m.logger.Info("remove stale service " + name + ".")
+		delete(m.services, name)
+	}
+}
+
+type service struct {
+	logger     *zap.Logger
+	name       string
+	cancelFunc context.CancelFunc
+
+	rwLock        sync.RWMutex
+	primaryCMS    *CountMinSketch
+	primaryTopK   *topKMetrics
+	secondaryCMS  *CountMinSketch
+	secondaryTopK *topKMetrics
+
+	totalCount    int
+	rotations     int
+	countSnapshot []int
+
+	totalRollup     int
+	totalMetricSent int
+}
+
+func (s *service) InsertMetricDataToPrimary(md *MetricData) {
+	s.primaryCMS.Insert(md)
+	updatedFrequency := s.primaryCMS.Get(md)
+	updatedMd := copyMetricDataWithUpdatedFrequency(md, updatedFrequency)
+	s.primaryTopK.Push(md, updatedMd)
+}
+
+func (s *service) InsertMetricDataToSecondary(md *MetricData) {
+	if s.secondaryCMS != nil {
+		s.secondaryCMS.Insert(md)
+		updatedFrequency := s.secondaryCMS.Get(md)
+		updatedMd := copyMetricDataWithUpdatedFrequency(md, updatedFrequency)
+		s.secondaryTopK.Push(md, updatedMd)
+	}
+}
+
+// MetricData represents a key-value pair.
+type MetricData struct {
+	labels    map[string]string
```

Comment:
not blocker: we can remove `labels` field in MetricData but just keep `hashKey` to reduce the memory usage. As MetricData will be stored as value for every service TopKMetrics map.